modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -190,6 +76,47 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def validateJoinConditions(): F[Unit] =
+    for {
+      nodeState <- nodeStorage.getNodeState
+      canJoinCluster <- nodeStorage.canJoinCluster
+      _ <- Applicative[F].unlessA(canJoinCluster)(NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit])
+    } yield ()
+
+  private def joinAndDiscover(context: P2PContext): F[Set[P2PContext]] = for {
```

Comment:
The return type of `joinAndDiscover` has been changed to `EitherT[F, Throwable, Set[P2PContext]]`. There is no point in differentiating from a joining or peer discovery failure, since no discovered peers, joining failure, or peer discovery failure will be treated the same, as `Set.empty[P2PContext]`. The failures will simply be logged. Both joining and peer discovery failures will return `Throwable` as `Left`.

Reply:
Well so you won't be able to differentiate between successful join with unsuccessful discovery from the unsuccessful join. I'm only pointing it out in terms of the joining functionality and some heuristic we may want to run at the end of joining process where we could determine whether the joining succeeded overall or didn't. For example joining successfully to a single node out of 100 attempted can be treated as unhealthy condition and may be handled with failing the joining process. Not saying we need to do that, but I see the opportunity and a benefit from such check.

Reply:
Unsuccessful joins return `FailedToJoin` while successful joins but unsuccessful peer discovery returns `FailedToDiscoverFrom`, so we can differentiate between the two cases.